src/transforms/sampler.rs

```diff
@@ -8,13 +8,22 @@ use regex::RegexSet; // TODO: use regex::bytes
 use serde::{Deserialize, Serialize};
 use snafu::ResultExt;
 
+#[derive(Deserialize, Serialize, Debug, Clone, Copy)]
+#[serde(rename_all = "snake_case")]
+pub enum SampleProperty {
+    Hash,
+    Index,
+}
+
 #[derive(Deserialize, Serialize, Debug, Clone)]
 #[serde(deny_unknown_fields)]
 pub struct SamplerConfig {
     pub rate: u64,
     pub key_field: Option<String>,
     #[serde(default)]
     pub pass_list: Vec<String>,
+    #[serde(default = "default_property")]
+    pub property: SampleProperty,
```

Comment:
I'm not sure this is something we need to expose, since it should be mutually exclusive with setting a `key_field`. Why don't we hash when a key is set and use the index when one is not set?

Reply:
That's a good option too. The main reason for exposing `property` is to enable usage of `pass_list` with index as it seams like a nice option to have. It also makes room for additional ways of sampling, like really random, or some other property.

Reply:
Why do we need to expose this for `pass_list` to work? I think we could just have `pass_list` be an initial check and not count events that match towards the total for index. It is essentially an opt-out of sampling for certain events.

I do see the benefit if we want to expose more ways to sample in the future, like `strategy` for file fingerprinting, but I'm not sure that we need it right now.

Reply:
Without exposing `property`, users can't specify on which filed should `pass_list` be applied to without also changing the sampling to hash.

Reply:
Ah, I see. It doesn't seem that `pass_list` is very useful if it must be an exact match of the `key_field`. Maybe we should change it to be a `Condition`?

Reply:
That sounds like a better approach. I'll change it.